idaes/core/util/performance.py

```diff
@@ -0,0 +1,169 @@
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES), and is copyright (c) 2018-2021
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+"""
+Author: Andrew Lee
+"""
+import gc
+import pytest
+
+from pyomo.util.check_units import assert_units_consistent
+from pyomo.common.timing import TicTocTimer
+from pyomo.environ import assert_optimal_termination
+import pyomo.common.unittest as unittest
+
+from idaes.core.solvers import get_solver
+from idaes.core.util.model_statistics import degrees_of_freedom
+
+# Get default solver for testing
+solver = get_solver()
+
+
+class PerformanceBaseClass:
+    """
+    Base class for running the IDAES Performance Test Suite.
+
+    This class is intended to be used as a basis for constructing tests to be run as
+    part of the IDAES performance testing suite, and should implement the following
+    methods:
+
+    * build_model - should return a complete instance of the model being tested.
+    * test_performance - a test method for profiling the model of interest which logs
+        some performance data. A standard method is provided which logs some common
+        performance data, but developers should overload this as required.
+    * initialize_model - (optional) a method to initialize the model of interest. This
+        method is called as part of the standard test_performance method.
+    * solve_model - (optional) a method to solve the model of interest. This method is
+        called as part of the standard test_performance method, and a standard method
+        is provided by the base class.
+
+    Developers may add additional test methods to the derived class to run additional
+    performance tests if desired. These methods should use standard pytest marks and
+    naming conventions.
+
+    Usage:
+
+    Performance tests should inherit from this based class and unittest.TestCase.
+    Inheritance from TestCase is used to make the derived classes easily discoverable
+    for use outside the performance testing infrastructure. The derived class should
+    be decorated with the 'performance' pytest mark. This will automatically categorize
+    all tests in this class as performance tests which will only be run when explicitly
+    called for.
+
+    Raises:
+        TypeError is derived class does not inherit from unittest.TestCase.
+    """
+
+    def __init_subclass__(cls):
+        if unittest.TestCase not in cls.mro():
```

Comment:
Note that this prevents creating a helper class that inherits from this, and from which multiple test suites then inherit.  That said, I don't thin it's a big deal, as the helper class can be implemented as a mixin class.

Reply:
My current plan is to implement a `DynamicPerformanceTest` subclass in order to override `test_performance` to (a) not check unit consistency and (b) potentially record the discretization transformation time as well. I would like to inherit from the base class so that I can still identify dynamic performance tests with `isinstance`. I think this restriction is enough that I would prefer this check to go in `__init__`.

Reply:
A couple of comments:
1. For dynamic models, hopefully Pyomo can get a fix in for the underlying problem, which would remove the need for this distinction.
2. I believe the transformation time should already be captured without the need for us to record it. Note that the Pyomo performance testing infrastructure logs a lot of information behind the scenes.

Reply:
@Robbybp, why would a mixin class not work?  You don't need `DynamicPerformanceTest` to inherit from `PerformanceBaseClass`, do you?  It should be sufficient for your dynamic tests to:
```python
class TestMyThing(DynamicPerformanceTest, PerformanceBaseClass, unittest.TestCase):
```

Reply:
Regardless, it would be useful to have an explicit distinction for dynamic models (and flowsheet models, and maybe other types of models as well) as we may want to benchmark different initialization or solution strategies that are only valid for these types of models.

Reply:
I am not sure if this is the best way to categorise things like that - I would suggest a set of special pytest marks would be better in this case (at least from the timing perspective at least). This would make it easy to run different subsets of the performance tests.

Secondly, I would also suggest that in at least some of these cases there should be multiple performance tests in a single test class (e.g. steady-state case, dynamic case, initialization method 1, initialization method 2, etc.).

Reply:
@jsiirola No reason that doesn't work, just didn't think of it and didn't know what a mixin class was. That is a little more complicated, but probably worth it for the benefit of failing at import time rather than runtime.

Reply:
I was guessing that was the intent - I just wonder if we should have some other way to distinguish things like that that we could use for multiple purposes.

Reply:
That sounds like Robby doesn't want different performance metrics, but rather he wants to collect diagnostics on performance test models?  These things (#vars, condition #, etc) are not expected to change over time, right?  It feels lie we are confounding performance tests with other test suite diagnostics...

Reply:
Yes, to an extent we are. I think we have 3 things here:

1. A standard suite of models we use for this type of testing.
2. A set of performance tests based on the model suite.
3. A set of diagnostics (#vars, etc.) on these models.

When we started 1 and 2 were intertwined, and we are trying to disentangle them now.

Reply:
Those assessments sound accurate. Maybe we need a list of "test models" (with an API for build/initialize/solve) somewhere, then, separately, a way to run a standard `test_performance` on each of them.